### PR TITLE
Provide structured suggestion for binding needing type on E0594

### PR DESCRIPTION
**Pull request by estebank**
Partially address #45405.

### CODE REVIEW

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 1]
```diff
@@ -606,12 +606,60 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+                                            if local.pat.span == self.span {
+                                                self.hir_id = Some(local.hir_id);
+                                            }
+                                        }
+                                        hir::intravisit::walk_stmt(self, s);
+                                    }
+                                }
+                                let hir_map = self.infcx.tcx.hir();
+                                let pat = loop {
+                                    // Poor man's try block
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:626]: Can we just use a try block here? This is the compiler after all. Or a closure `(|| {})()` is typically what I see being used instead.

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 2]
```diff
@@ -606,12 +606,60 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+                                    let def_id = self.body.source.def_id();
+                                    let hir_id =
+                                        hir_map.local_def_id_to_hir_id(def_id.as_local().unwrap());
+                                    let node = hir_map.find(hir_id);
+                                    let Some(hir::Node::Item(item)) = node else { break None; };
+                                    let hir::ItemKind::Fn(.., body_id) = item.kind else { break None; };
+                                    let body = self.infcx.tcx.hir().body(body_id);
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:633]:
```suggestion
                                    let body = hir_map.body(body_id);
```

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 3]
```diff
@@ -606,12 +606,60 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+                                        }
+                                        hir::intravisit::walk_stmt(self, s);
+                                    }
+                                }
+                                let hir_map = self.infcx.tcx.hir();
+                                let pat = loop {
+                                    // Poor man's try block
+                                    let def_id = self.body.source.def_id();
+                                    let hir_id =
+                                        hir_map.local_def_id_to_hir_id(def_id.as_local().unwrap());
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:629]:
```suggestion
                                        hir_map.local_def_id_to_hir_id(def_id.expect_local());
```

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 4]
```diff
@@ -606,12 +606,60 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                                 }
                             }
                             Some((false, err_label_span, message)) => {
-                                err.span_label(
-                                    err_label_span,
-                                    &format!(
-                                        "consider changing this binding's type to be: `{message}`"
-                                    ),
-                                );
+                                struct V {
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:609]: This deserves a better name, I think.

**Comment by compiler-errors** [comment]
r=me after the nits tho

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:626]: `#![feature(try_blocks)]` is already enabled in `rustc_borrowck` too :smiley:

Comment (estebank) [reply compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs]: Ended up doing let chains instead :)

**Comment by compiler-errors** [comment]
thanks

@bors r+ rollup

**Pull request merged**